### PR TITLE
Python print

### PR DESCRIPTION
**Pull request by Rombur**
Add functions to print `LA::Vector` and `SparseMatrix` on a format than can be read by numpy

### CODE REVIEW

**Comment by masterleinad** [comment]
The indentation is off.

File: include/deal.II/lac/sparse_matrix.h [hunk 1]
```diff
@@ -1583,6 +1583,17 @@ class SparseMatrix : public virtual Subscriptor
   void
   print_pattern(std::ostream &out, const double threshold = 0.) const;
 
+  /**
+   * Print the vector to the output stream @p out in a format than can be
```
Comment (masterleinad) [inline include/deal.II/lac/sparse_matrix.h]:
```suggestion
   * Print the vector to the output stream @p out in a format that can be
```

File: tests/lac/la_vector_print.cc [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+
+int
+main()
+{
+  std::ofstream logfile("output");
+  deallog << std::fixed;
+  deallog << std::setprecision(2);
+  deallog.attach(logfile);
```
Comment (masterleinad) [inline tests/lac/la_vector_print.cc]: Can't you just do `initlog()`?

File: tests/lac/sparse_matrix_print.cc [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+
+int
+main()
+{
+  std::ofstream logfile("output");
+  deallog << std::fixed;
+  deallog << std::setprecision(2);
+  deallog.attach(logfile);
```
Comment (masterleinad) [inline tests/lac/sparse_matrix_print.cc]: Can't you just do `initlog()`?

File: include/deal.II/lac/sparse_matrix.templates.h [hunk 4]
```diff
@@ -1965,6 +1965,50 @@ SparseMatrix<number>::print_pattern(std::ostream &out,
...
+        {
+          rows.push_back(i);
+          columns.push_back(cols->colnums[j]);
+          values.push_back(val[j]);
+        }
+    }
+
+  for (auto d : values)
+    out << d << " ";
+  out << "\n";
```
Comment (masterleinad) [inline include/deal.II/lac/sparse_matrix.templates.h]:
```suggestion
  out << '\n';
```

File: include/deal.II/lac/sparse_matrix.templates.h [hunk 5]
```diff
@@ -1965,6 +1965,50 @@ SparseMatrix<number>::print_pattern(std::ostream &out,
...
+
+  for (auto r : rows)
+    out << r << " ";
+  out << "\n";
```
Comment (masterleinad) [inline include/deal.II/lac/sparse_matrix.templates.h]:
```suggestion
  out << '\n';
```

File: include/deal.II/lac/sparse_matrix.templates.h [hunk 6]
```diff
@@ -1965,6 +1965,50 @@ SparseMatrix<number>::print_pattern(std::ostream &out,
...
+
+  for (auto c : columns)
+    out << c << " ";
+  out << "\n";
```
Comment (masterleinad) [inline include/deal.II/lac/sparse_matrix.templates.h]:
```suggestion
  out << '\n';
```

File: include/deal.II/lac/sparse_matrix.templates.h [hunk 7]
```diff
@@ -1965,6 +1965,50 @@ SparseMatrix<number>::print_pattern(std::ostream &out,
...
+  for (auto d : values)
+    out << d << " ";
+  out << "\n";
+
+  for (auto r : rows)
+    out << r << " ";
+  out << "\n";
+
+  for (auto c : columns)
+    out << c << " ";
```
Comment (masterleinad) [inline include/deal.II/lac/sparse_matrix.templates.h]:
```suggestion
    out << c << ' ';
```

File: include/deal.II/lac/sparse_matrix.templates.h [hunk 8]
```diff
@@ -1965,6 +1965,50 @@ SparseMatrix<number>::print_pattern(std::ostream &out,
...
+          values.push_back(val[j]);
+        }
+    }
+
+  for (auto d : values)
+    out << d << " ";
+  out << "\n";
+
+  for (auto r : rows)
+    out << r << " ";
```
Comment (masterleinad) [inline include/deal.II/lac/sparse_matrix.templates.h]:
```suggestion
    out << r << ' ';
```

**Comment by dangars** [comment]
@Rombur Wouldn't it be better to add the Containers `LA::Vector` and `SparseMatrix` to the `HDF5` interface? I can write the HDF5 interfaces. I had in my todo list to add the `LA::Vector` container to the `HDF5` interface.

I wrote the `HDF5` interface in such a way that it is easy to add new Containers and types.

If you use `numpy.loadtxt` then you need a machine that has enough ram memory to store the whole dataset in memory. This could be an issue if you have a very large simulation. That is not a problem with `h5py` and `HDF5`. You only need enough memory for the slice that you want to read.

It would be possible to have both options in the library: this interface and the `HDF5` interface.

The `LA::Vector` container should be straightforward. There can be two options depending on the usual size of `IndexSet`:
- `LA::Vector` is a `HDF5::DataSet` and `IndexSet` is an attribute of the `DataSet`. Attributes are supposed to be small. Attributes can not be chunked.
- `LA::Vector` is a `HDF5::Group` in which the you have two `DataSets`, one for the `Vector` data and another one for the `IndexSet`

Regarding `SparseMatrix`. `HDF5` does not provide any sparse matrix built-in type. The best option would be to store `SparseMatrix` as a `HDF5::Group`. If we use something like Compressed Sparse Row matrix (csr), then the data can be easily read by `scipy.sparse` using `h5py`.

Let me know if you find that this can be useful and if I should write `HDF5` interfaces for `LA::Vector` and `SparseMatrix` :)

**Comment by Rombur** [comment]
@dangars Well it depends. I don't have hdf5 installed on my machine so for me it's not better :p I use these functions to help me debug so the matrices are always very small. Usually, I just want to check that the matrix SPD and that can be done on a coarse mesh. But I think that it could be useful to also have the `hdf5` back end. Something I'd like to do is to add python wrappers to some of the deal.II solvers and preconditioners. My goal is to be able to dump a matrix from C++, load it in python, and then try different solvers/preconditioners from deal.II and scipy to see what works best.

File: include/deal.II/lac/sparse_matrix.templates.h [hunk 9]
```diff
@@ -1965,6 +1965,50 @@ SparseMatrix<number>::print_pattern(std::ostream &out,
...
 
+template <typename number>
+void
+SparseMatrix<number>::print_numpy_array(std::ostream &     out,
+                                        const unsigned int precision) const
+{
+  Assert(cols != nullptr, ExcNotInitialized());
+  Assert(val != nullptr, ExcNotInitialized());
+
+  bool                hanging_diagonal = false;
```
Comment (masterleinad) [inline include/deal.II/lac/sparse_matrix.templates.h]: This variable is unused.

File: include/deal.II/lac/la_vector.templates.h [hunk 10]
```diff
@@ -560,6 +560,25 @@ namespace LinearAlgebra
...
 
 
+  template <typename Number>
+  void
+  Vector<Number>::print_numpy_array(std::ostream &     out,
+                                    const unsigned int precision) const
+  {
+    AssertThrow(out, ExcIO());
+
+    out.precision(precision);
```
Comment (masterleinad) [inline include/deal.II/lac/la_vector.templates.h]: We should probably restore the previous precision after calling this function.

Comment (tjhei) [reply include/deal.II/lac/la_vector.templates.h]: this can be done using ``boost::io::ios_flags_saver`` for example.

**Comment by dangars** [comment]
@Rombur That makes sense, it is good to have both interfaces :)

I will prepare a first version of the `hdf5` interface and let you know when it is ready.

Regarding the python wrappers to some of the deal.II solvers. I see two approaches:
- Exchange the data between python and deal.ii using your interface or `hdf5`. This should be easy to implement once the `Vector` and `SparceMatrix` interfaces are ready. There can be an option to choose the back end `hdf5` or `print_numpy_array`.
- Use C bindings. It would be hard to implement with solvers that use MPI.

I discuss some ideas about the python bindings in #2914

**Comment by Rombur** [comment]
I have done the changes asked.

**Comment by masterleinad** [comment]
It seems that you are missing a header here:
```
In file included from /jenkins/workspace/dealii_PR-8299/source/lac/la_vector.cc:16:0:
/jenkins/workspace/dealii_PR-8299/include/deal.II/lac/la_vector.templates.h: In member function ‘void dealii::LinearAlgebra::Vector<Number>::print_numpy_array(std::ostream&, unsigned int) const’:
/jenkins/workspace/dealii_PR-8299/include/deal.II/lac/la_vector.templates.h:569:12: error: ‘boost::io’ has not been declared
     boost::io::ios_flags_saver restore_flags(out);
            ^
````

**Comment by masterleinad** [comment]
I just pushed the missing header myself.

File: include/deal.II/lac/la_vector.templates.h [hunk 11]
```diff
@@ -560,6 +562,26 @@ namespace LinearAlgebra
...
+  {
+    AssertThrow(out, ExcIO());
+    boost::io::ios_flags_saver restore_flags(out);
+
+    out.precision(precision);
+
+    const unsigned int n_elements = this->n_elements();
+    for (unsigned int i = 0; i < n_elements; ++i)
+      out << this->values[i] << ' ';
+    out << '\n' << std::flush;
```
Comment (bangerth) [inline include/deal.II/lac/la_vector.templates.h]: Completely unimportant, but for future reference: The sequence `<< '\n' << std::flush` is precisely how `<< std::endl` is defined.

File: include/deal.II/lac/sparse_matrix.h [hunk 12]
```diff
@@ -1583,6 +1583,17 @@ class SparseMatrix : public virtual Subscriptor
   void
   print_pattern(std::ostream &out, const double threshold = 0.) const;
 
+  /**
+   * Print the vector to the output stream @p out in a format that can be
```
Comment (bangerth) [inline include/deal.II/lac/sparse_matrix.h]: It's not a vector :-(

File: include/deal.II/lac/la_vector.h [hunk 13]
```diff
@@ -344,14 +344,26 @@ namespace LinearAlgebra
...
+    /**
+     * Print the vector to the output stream @p out in a format that can be
+     * read by numpy::readtxt(). Note that the IndexSet is not printed but only
+     * the values stored in the Vector. To load the vector in python just do
+     * <code>
+     * vector = numpy.loadtxt('my_vector.txt')
+     * </code>
+     */
+    void
+    print_numpy_array(std::ostream &     out,
```
Comment (bangerth) [inline include/deal.II/lac/la_vector.h]: Would you be willing to name this function `print_as_numpy_array()`? I think that makes the intent clearer.

File: include/deal.II/lac/sparse_matrix.h [hunk 14]
```diff
@@ -1583,6 +1583,17 @@ class SparseMatrix : public virtual Subscriptor
...
+   * read by numpy::readtxt(). To load the matrix in python just do
+   * <code>
+   *  [data, row, column] = numpy.loadtxt('my_matrix.txt')
+   *  sparse_matrix = scipy.sparse.csr_matrix((data, (row, column)))
+   * </code>
+   */
+  void
+  print_numpy_array(std::ostream &out, const unsigned int precision = 9) const;
```
Comment (bangerth) [inline include/deal.II/lac/sparse_matrix.h]: same as above, but here I would also use the plural: `print_as_numpy_arrays()`.

File: doc/news/changes/minor/20180602BrunoTurcksin [hunk 15]
```diff
@@ -0,0 +1,4 @@
+New: Add function print_numpy_array to SparseMatrix and LinearAlgebra::Vector to
```
Comment (bangerth) [inline doc/news/changes/minor/20180602BrunoTurcksin]: If you write it as follows, then doxygen will automatically link to the new function:
```suggestion
New: Add function to SparseMatrix::print_numpy_array() and LinearAlgebra::Vector::print_numpy_array() to
```
But see the note about function names below.

**Pull request merged**